panic: runtime error: invalid memory address or nil pointer dereference

```
$ cat redis.go
package main

import rds "gopkg.in/redis.v3"

func main() {
	client := rds.NewClient(&rds.Options{
		// Network: "tcp",
		Addr: "192.168.33.20:6379",
	})
	if client == nil {
		panic("Redis NewClient failed")
	}

	err := client.Ping().Err()

	if err != nil {
		panic("rds.NewClient error : " + err.Error())
	}
}
```

```
$ go run redis.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x80dec6c]

goroutine 1 [running]:
panic(0x81c0000, 0x1850a038)
	/home/vagrant/dist/go/src/runtime/panic.go:464 +0x326
sync/atomic.AddUint64(0x18576104, 0x1, 0x0, 0x0, 0x184fe7de)
	/home/vagrant/dist/go/src/sync/atomic/asm_386.s:112 +0xc
gopkg.in/redis%2ev3.(*connPool).Get(0x185760f0, 0x0, 0x82df100, 0x0, 0x0)
	/tgo/src/gopkg.in/redis.v3/pool.go:240 +0x8a
gopkg.in/redis%2ev3.(*baseClient).conn(0x1850e350, 0x1, 0x185000d0, 0x0, 0x0)
	/tgo/src/gopkg.in/redis.v3/redis.go:23 +0x3d
gopkg.in/redis%2ev3.(*baseClient).process(0x1850e350, 0xb6c11210, 0x18518210)
	/tgo/src/gopkg.in/redis.v3/redis.go:43 +0x5c
gopkg.in/redis%2ev3.(*baseClient).(gopkg.in/redis%2ev3.process)-fm(0xb6c11210, 0x18518210)
	/tgo/src/gopkg.in/redis.v3/redis.go:189 +0x30
gopkg.in/redis%2ev3.(*commandable).Process(0x1850a21c, 0xb6c11210, 0x18518210)
	/tgo/src/gopkg.in/redis.v3/commands.go:57 +0x2e
gopkg.in/redis%2ev3.(*commandable).Ping(0x1850a21c, 0x1850a218)
	/tgo/src/gopkg.in/redis.v3/commands.go:77 +0xed
main.main()
	/tgo/src/twibo/redis.go:16 +0xb7
exit status 2
```

```
$ go version
go version go1.6 linux/386
```


Am I doing anything wrong?

--- a/pool.go
+++ b/pool.go
@@ -17,12 +17,12 @@ var (
 
 // PoolStats contains pool state information and accumulated stats.
 type PoolStats struct {
-	Requests uint64 // number of times a connection was requested by the pool
-	Waits    uint64 // number of times our pool had to wait for a connection
-	Timeouts uint64 // number of times a wait timeout occurred
+	Requests uint32 // number of times a connection was requested by the pool
+	Waits    uint32 // number of times our pool had to wait for a connection
+	Timeouts uint32 // number of times a wait timeout occurred
 
-	TotalConns uint64 // the number of total connections in the pool
-	FreeConns  uint64 // the number of free connections in the pool
+	TotalConns uint32 // the number of total connections in the pool
+	FreeConns  uint32 // the number of free connections in the pool
 }
 
 type pool interface {
@@ -237,7 +237,7 @@ func (p *connPool) Get() (cn *conn, isNew bool, err error) {
 		return
 	}
 
-	atomic.AddUint64(&p.stats.Requests, 1)
+	atomic.AddUint32(&p.stats.Requests, 1)
 
 	// Fetch first non-idle connection, if available.
 	if cn = p.First(); cn != nil {
@@ -257,12 +257,12 @@ func (p *connPool) Get() (cn *conn, isNew bool, err error) {
 	}
 
 	// Otherwise, wait for the available connection.
-	atomic.AddUint64(&p.stats.Waits, 1)
+	atomic.AddUint32(&p.stats.Waits, 1)
 	if cn = p.wait(); cn != nil {
 		return
 	}
 
-	atomic.AddUint64(&p.stats.Timeouts, 1)
+	atomic.AddUint32(&p.stats.Timeouts, 1)
 	err = errPoolTimeout
 	return
 }
@@ -315,11 +315,11 @@ func (p *connPool) FreeLen() int {
 
 func (p *connPool) Stats() *PoolStats {
 	stats := p.stats
-	stats.Requests = atomic.LoadUint64(&p.stats.Requests)
-	stats.Waits = atomic.LoadUint64(&p.stats.Waits)
-	stats.Timeouts = atomic.LoadUint64(&p.stats.Timeouts)
-	stats.TotalConns = uint64(p.Len())
-	stats.FreeConns = uint64(p.FreeLen())
+	stats.Requests = atomic.LoadUint32(&p.stats.Requests)
+	stats.Waits = atomic.LoadUint32(&p.stats.Waits)
+	stats.Timeouts = atomic.LoadUint32(&p.stats.Timeouts)
+	stats.TotalConns = uint32(p.Len())
+	stats.FreeConns = uint32(p.FreeLen())
 	return &stats
 }
 

